File: LittleProjects/Scala/AkkaInfinispanCrud/src/main/scala/akka/infinispan/crud/repository/EntityRepository.scala
package akka.infinispan.crud.repository

import java.util.stream.Collectors

import akka.infinispan.crud.common.{CacheEventListener, DataSource}
import akka.infinispan.crud.model.Entity
import cats.effect.IO
import javax.transaction.TransactionManager
import org.infinispan.Cache
import org.infinispan.context.Flag

import scala.collection.JavaConverters._

class EntityRepository extends Repository[Long, Entity] {
  lazy val cache: Cache[Long, Entity] = {
    val c: Cache[Long, Entity] = DataSource.cacheManager.getCache("entity")
    c.addListener(CacheEventListener())
    c
  }

  override def findOne(id: Long): IO[Option[Entity]] = IO {
    cache.get(id) match {
      case null => None
      case value => Some(value)
    }
  }

  override def findAll(): IO[Seq[Entity]] = IO(cache.values().stream().collect(Collectors.toList[Entity]).asScala)

  override def delete(id: Long): IO[Unit] = IO(cache.remove(id))

  override def deleteAll(): IO[Unit] = IO(cache.entrySet().forEach(e => cache.remove(e.getKey)))

  override def save(a: Entity): IO[Unit] = IO(cache.put(a.id, a))

  override def update(a: Entity): IO[Unit] = IO(cache.replace(a.id, a))

  override def transaction: TransactionManager = cache.getAdvancedCache.getTransactionManager
}

object EntityRepository {
  def apply(): EntityRepository = new EntityRepository()
}

File: LittleProjects/Java/GuiceHibernatePostgresCrud/src/main/java/guice/hibernate/postgres/crud/repository/EntityRepository.java
package guice.hibernate.postgres.crud.repository;

import guice.hibernate.postgres.crud.model.Entity;

import javax.persistence.EntityManager;
import java.util.Optional;

public class EntityRepository implements Repository<Entity, Long> {
    public EntityManager em;

    public EntityRepository() {
    }

    public EntityRepository(EntityManager em) {
        this.em = em;
    }

    @Override
    public Optional<Entity> findOne(Long aLong) {
        return Optional.of(em.find(Entity.class, aLong));
    }

    @Override
    public Iterable<Entity> findAll() {
        return em.createNamedQuery("ALL_ENTITIES", Entity.class).getResultList();
    }

    @Override
    public Entity save(Entity entity) {
        em.persist(entity);
        return entity;
    }

    @Override
    public Entity update(Entity entity) {
        em.persist(entity);
        return entity;
    }

    @Override
    public boolean deleteOne(Long aLong) {
        return em.createNamedQuery("DELETE_ENTITY_BY_ID").setParameter("id", aLong).executeUpdate() != 0;
    }

    @Override
    public long deleteAll() {
        return em.createNamedQuery("DELETE_ALL_ENTITIES").executeUpdate();
    }
}
